File: cargo/crates/numbers-utils/src/operations/factorial_utils.rs
use num_bigint::BigUint;
use num_bigint::ToBigUint;
use rand::Rng;

/// # `factorial`
///
/// This function calculates the factorial of a number.
///
/// __Arguments:__
///
/// * `num: usize` - The number to calculate the factorial of.
///
/// __Return:__
///
/// * `BigUint` - The factorial of the number `num`.
///
/// __Example:__
/// ```rust
/// use numbers_utils::operations::factorial_utils::factorial;
/// let fact = factorial(5);
/// println!("5! = {}", fact);
/// ```
///
/// __Output:__
/// ```txt
/// 5! = 120
/// ```
pub fn factorial(num: usize) -> BigUint {
    let mut fact = BigUint::from(1u8);

    for i in 1..=num {
        fact *= i.to_biguint().unwrap();
    }

    fact
}

/// # `factorial_random`
///
/// This function generates a random number and returns its factorial.
///
/// __Return:__
///
/// * `String` - The factorial of the random number.
///
/// __Example:__
/// ```rust
/// use numbers_utils::operations::factorial_utils::factorial_random;
/// let fact = factorial_random();
/// println!("{}", fact);
/// ```
///
/// __Output:__
/// ```txt
/// Fact[5] = 120
/// ```
pub fn factorial_random() -> String {
    let mut rng = rand::thread_rng();
    let random_num = rng.gen_range(1..=25);

    let fact = factorial(random_num as usize);

    format!("Fact[{}] = {}", random_num, fact)
}

File: kotlin/subprojects/numbers-utils/src/main/kotlin/com/prog/numbers/utils/operations/FactorialUtils.kt
package com.prog.numbers.utils.operations

import java.math.BigInteger
import kotlin.random.Random

object FactorialUtils {
    /**
     * # `factorial`
     *
     * This function calculates the factorial of a given number.
     *
     * __Arguments:__
     *
     * * `n: Int` - The number to calculate the factorial
     *
     * __Return:__
     *
     * * `BigInteger` - The factorial of the given number
     *
     * __Example:__
     *
     * ```kotlin
     * import com.prog.numbers.utils.operations.FactorialUtils.factorial
     * val fact = factorial(5)
     * println("Fact[5] = $fact")
     * ```
     *
     * __Output:__
     *
     * ```kotlin
     * "Fact[5] = 120"
     * ```
     */
    fun factorial(n: Int): BigInteger {
        var fact = BigInteger.ONE

        for (i in 1..n) {
            fact *= BigInteger.valueOf(i.toLong())
        }

        return fact
    }

    /**
     * # `factorialRandom`
     *
     * This function generates a random number and calculates its factorial.
     *
     * __Return:__
     *
     * * `String` - The factorial of the random number
     *
     * __Example:__
     *
     * ```kotlin
     * import com.prog.numbers.utils.operations.FactorialUtils.factorialRandom
     * val fact = factorialRandom()
     * println(fact)
     * ```
     *
     * __Output:__
     *
     * ```kotlin
     * "Fact[5] = 120"
     * ```
     */
    fun factorialRandom(): String {
        val rng = Random
        val randomNum = rng.nextInt(1, 26)
        val fact = factorial(randomNum)

        return "Fact[$randomNum] = $fact"
    }
}
